### PR TITLE
Fix tests for change in PostgreSQL 14 behavior change.

### LINKED ISSUE #14298: Trial tests against postgres 14 are failing since #11635
Latest deps: https://github.com/matrix-org/synapse/actions/runs/3326877285/jobs/5501039194
Develop: https://github.com/matrix-org/synapse/actions/runs/3323214952/jobs/5493346783

```
[FAIL]
Traceback (most recent call last):
  File "/home/runner/work/synapse/synapse/tests/storage/test_room_search.py", line 372, in test_postgres_web_search_for_phrase
    self._check_test_cases(store, self.COMMON_CASES + self.POSTGRES_CASES, index=0)
  File "/home/runner/work/synapse/synapse/tests/storage/test_room_search.py", line 323, in _check_test_cases
    self.assertEquals(
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.10/lib/python3.10/site-packages/twisted/trial/_synctest.py", line 424, in assertEqual
    super().assertEqual(first, second, msg)
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/unittest/case.py", line 845, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/unittest/case.py", line 838, in _baseAssertEqual
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: 0 != 1 : expected '"fox quick' to match 'the quick brown fox jumps over the lazy dog'

tests.storage.test_room_search.MessageSearchTest.test_postgres_web_search_for_phrase
```

From #11635.

### PR DESCRIPTION
**Pull request by clokep**
See #14298 for details, but postgresql 14 changed the behavior of `websearch_to_tsquery` to fix some bugs. The tests were hitting those edge-cases though.

This is a bit ugly, but I couldn't come up with a better way to do it...

Fixes #14298.

This was broken in #11635.

### CODE REVIEW

File: synapse/storage/databases/main/search.py [hunk 1]
```diff
@@ -824,9 +824,8 @@ def _tokenize_query(query: str) -> TokenList:
     in_phrase = False
     parts = deque(query.split('"'))
     for i, part in enumerate(parts):
-        # The contents inside double quotes is treated as a phrase, a trailing
-        # double quote is not implied.
-        in_phrase = bool(i % 2) and i != (len(parts) - 1)
+        # The contents inside double quotes is treated as a phrase.
+        in_phrase = bool(i % 2)
```
Comment (DMRobertson) [inline synapse/storage/databases/main/search.py:828]: To check: this means that a trailing quote _is_ now implied?

File: tests/storage/test_room_search.py [hunk 2]
```diff
@@ -269,6 +268,14 @@ def prepare(
         response = self.helper.send(self.room_id, self.PHRASE, tok=self.access_token)
         self.assertIn("event_id", response)
 
+        # The behaviour of a missing trailing double quote changed in PostgreSQL 14
+        # from ignoring the initial double quote to treating it as a phrase.
```
Comment (DMRobertson) [inline tests/storage/test_room_search.py:272]: To check: we've changed our parser to mirror this behaviour?

File: synapse/storage/databases/main/search.py (hunk 1 above)
Comment (clokep) [reply synapse/storage/databases/main/search.py:828]: Yes. 👍

File: tests/storage/test_room_search.py (hunk 2 above)
Comment (clokep) [reply tests/storage/test_room_search.py:272]: Yes, so SQLite will match the behavior of PostgreSQL 14. PostgreSQL 11 - 13 will still use whatever their internal function call does. (And PostgreSQL 10 fallsback to `plainto_tsquery` anyway so let's just ignore talking about that.)

**Review by DMRobertson — APPROVED** [review]
Thanks, LGTM!

File: tests/storage/test_room_search.py [hunk 3]
```diff
@@ -269,6 +268,14 @@ def prepare(
...
+        main_store = homeserver.get_datastores().main
+        found = False
+        if isinstance(main_store.database_engine, PostgresEngine):
+            found = main_store.database_engine._version < 140000
+        self.COMMON_CASES.append(('"fox quick', (found, True)))
```
[the remark this answers is not in the document]
Comment (clokep) [reply tests/storage/test_room_search.py:277]: Not really, what it means is that:

* `fox` and `quick` are both found on PG < 14
* The phrase "fox quick" is not found on PG >= 14 and SQLite.

**Comment by DMRobertson** [comment]
> Thanks, LGTM!

... err, once CI is happy.

```
Run poetry run mypy
tests/storage/test_room_search.py:276: error: Unsupported operand types for > ("int" and "None")  [operator]
tests/storage/test_room_search.py:276: note: Left operand is of type "Optional[int]"
Found 1 error in 1 file (checked 738 source files)
Error: Process completed with exit code 1.
```

**Review by DMRobertson — APPROVED** [review]
Thanks, I'm happy if CI is!
